### PR TITLE
[XM] Add partial static registrar support

### PR DESCRIPTION
**Pull request by chamons**
This PR adds partial static registrar support for XM. For simple projects, it can dramatically improve application launch time.

A quick benchmark of NSOutliveViewAndTableExample showed Main to AppDelegate finish launching times from 3124 ms down to 349 ms.

I believe the change seems mostly safe for a C10 feature, but would like a serious review, as this isn't my normal area.

### CODE REVIEW

File: tools/mmp/Makefile [hunk 1]
```diff
@@ -150,9 +151,30 @@ $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/Mono.Cecil.dll: $(MONO_CECIL_
...
+	$(Q_GEN) $(SYSTEM_MONO) --debug mmp.exe --xamarin-framework-directory=$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR) -q --runregistrar:$(abspath $@) --sdkroot $(XCODE_DEVELOPER_ROOT) --sdk $(OSX_SDK_VERSION)   $< --registrar:static --target-framework Xamarin.Mac,v1.0 --arch=i386
+	$(Q) touch Xamarin.Mac.registrar.mac.i386.m Xamarin.Mac.registrar.mac.i386.h
+
+Xamarin.Mac.registrar.mac.x86_64.m:     $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/Xamarin.Mac/Xamarin.Mac.dll mmp.exe
+	$(Q_GEN) $(SYSTEM_MONO) --debug mmp.exe --xamarin-framework-directory=$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR) -q --runregistrar:$(abspath $@) --sdkroot $(XCODE_DEVELOPER_ROOT) --sdk $(OSX_SDK_VERSION)   $< --registrar:static --target-framework Xamarin.Mac,v1.0 --arch=x86_64
+	$(Q) touch Xamarin.Mac.registrar.mac.x86_64.m Xamarin.Mac.registrar.mac.x86_64.h
+
+Xamarin.Mac.registrar.mac.i386.a:   Xamarin.Mac.registrar.mac.i386.m
+	$(Q_CC) xcrun -sdk macosx clang -DDEBUG -g -gdwarf-2 -x objective-c++ -o $@ -c -arch i386 $< -Wall -I$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/include -mmacosx-version-min=10.7 -fobjc-runtime=macosx
```
Comment (rolfbjarne) [inline tools/mmp/Makefile]: Use `$(MIN_OSX_SDK_VERSION)` instead of hardcoding `10.7`.

File: tools/mmp/Makefile [hunk 2]
```diff
@@ -150,9 +151,30 @@ $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/Mono.Cecil.dll: $(MONO_CECIL_
...
+
+Xamarin.Mac.registrar.mac.x86_64.a: Xamarin.Mac.registrar.mac.x86_64.m
+	$(Q_CC) xcrun -sdk macosx clang -DDEBUG -g -gdwarf-2 -x objective-c++ -o $@ -c -arch x86_64 $< -Wall -I$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/include -mmacosx-version-min=10.7 -fobjc-runtime=macosx
```
Comment (rolfbjarne) [inline tools/mmp/Makefile]: Use $(MIN_OSX_SDK_VERSION) instead of hardcoding 10.7.

File: tools/mmp/Makefile [hunk 3]
```diff
@@ -132,6 +132,7 @@ MMP_TARGETS = \
 	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/mmp.exe \
 	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/Mono.Cecil.dll \
 	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/Mono.Cecil.Mdb.dll \
+	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/Xamarin.Mac.registrar.mac.a \
```
Comment (rolfbjarne) [inline tools/mmp/Makefile]: You'll need one static library per version of Xamarin.Mac.dll (which means one for mobile, one for 4.5 and one for full). This is because the generated code depends on metadata tokens, which is something that changes between the dlls.

File: tools/common/StaticRegistrar.cs [hunk 4]
```diff
@@ -642,6 +642,10 @@ protected override bool HasThisAttribute (MethodDefinition method)
 			
 		protected override bool Is64Bits {
 			get {
+#if !MTOUCH
```
Comment (rolfbjarne) [inline tools/common/StaticRegistrar.cs]: Instead of adding this condition, you can remove the `IsSingleAssembly` check below (it's not true anymore that we can't emit size-specific code in single assembly mode).

File: tools/mmp/driver.cs [hunk 5]
```diff
@@ -110,6 +110,7 @@ public static partial class Driver {
 
 		static bool is_extension;
 		static bool frameworks_copied_to_bundle_dir;	// Have we copied any frameworks to Foo.app/Contents/Frameworks?
+		static bool just_run_registrar;
```
Comment (rolfbjarne) [inline tools/mmp/driver.cs]: Just for the same of making the code more similar to mtouch IMHO it would be nicer to copy that code (i.e. have an `Action` enum (with just two entries in this case: Build and RunRegistrar) and an `action` variable that specifies what to do).

**Comment by chamons** [comment]
Change's aren't complete, I need to teach mmp to use the full/mobile flavor of lib and test.

**Comment by chamons** [comment]
@rolfbjarne That should be all of the requested changes. Please double check my Makefile. I _think_ i got it right, but Make is crazy.

**Comment by chamons** [comment]
Looks like a real failure. Let me dig into this...

**Comment by chamons** [comment]
Ok. Looks like parallel build is not working right at all. Hangs locally.

**Comment by chamons** [comment]
Now just test failures. Look unrelated.

**Comment by chamons** [comment]
build

**Comment by rolfbjarne** [comment]
This failure repeated and might be because of this PR:

```
1) CIFilterOutputImageTest (Xamarin.Mac.Tests.CIFilterTests.CIFilterOutputImageTest)
     CIFilterOutputImageTest - output was null
  Expected: not null
  But was:  null

  at Xamarin.Mac.Tests.CIFilterTests.CIFilterOutputImageTest () [0x0008a] in /Users/builder/jenkins/workspace/xamarin-macios-pr-builder/tests/apitest/src/CoreImage/CIFilter.cs:39 
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /Users/builder/jenkins/workspace/xamarin-macios-pr-builder/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/git/src/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305 
```

The rest are unrelated:
* Crash in apitest/Classic (bug #[46116](https://bugzilla.xamarin.com/show_bug.cgi?id=46116))
* Build failure in link all/watchOS (bug #[46097](https://bugzilla.xamarin.com/show_bug.cgi?id=46097))

**Comment by rolfbjarne** [comment]
build

**Comment by chamons** [comment]
Ok. We're down to "sane" test failures.

But given the previous mess, I would like to write a specific test to make sure we don't regress XM 4.5 + registrar again"...

File: tools/mmp/Makefile [hunk 6]
```diff
@@ -150,9 +152,40 @@ $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mmp/Mono.Cecil.dll: $(MONO_CECIL_
...
+
+Xamarin.Mac.registrar.mobile.i386.m:     $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/Xamarin.Mac/Xamarin.Mac.dll mmp.exe
+	$(GENERATE_PART_REGISTRAR) --target-framework Xamarin.Mac,v1.0 --arch=i386
+	$(Q) touch Xamarin.Mac.registrar.mobile.i386.m Xamarin.Mac.registrar.mobile.i386.h
+
+Xamarin.Mac.registrar.mobile.x86_64.m:     $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/Xamarin.Mac/Xamarin.Mac.dll mmp.exe
+	$(GENERATE_PART_REGISTRAR) --target-framework Xamarin.Mac,v1.0 --arch=x86_64
+	$(Q) touch Xamarin.Mac.registrar.mobile.x86_64.m Xamarin.Mac.registrar.mobile.x86_64.h
+
+Xamarin.Mac.registrar.full.i386.m:     $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/4.5/Xamarin.Mac.dll mmp.exe
```
Comment (rolfbjarne) [inline tools/mmp/Makefile]: Will this end up processing the correct 32-bit/64-bit version of Xamarin.Mac.dll?

File: tools/mmp/driver.cs [hunk 7]
```diff
@@ -148,8 +156,40 @@ public static partial class Driver {
 
 		public static Version MinOSVersion { get { return minos; } }
 
+		public static string ProductAssembly => "Xamarin.Mac";
+		public static string PlatformFrameworkDirectory
```
Comment (spouliot) [inline tools/mmp/driver.cs]: style `{` should be on the same line for properties

File: tools/mmp/driver.cs [hunk 8]
```diff
@@ -148,8 +156,40 @@ public static partial class Driver {
...
+		{
+			get
+			{
```
Comment (spouliot) [inline tools/mmp/driver.cs]: style `{` should be on the same line for properties

File: tools/mmp/driver.cs [hunk 9]
```diff
@@ -397,13 +446,23 @@ static void Main2 (string [] args)
...
+
 			App.InitializeCommon ();
 
 			Log ("Xamarin.Mac {0}{1}", Constants.Version, verbose > 0 ? "." + Constants.Revision : string.Empty);
 
 			if (verbose > 0)
 				Console.WriteLine ("Selected target framework: {0}; API: {1}", targetFramework, IsClassic ? "Classic" : "Unified");
 
+			if (action == Action.RunRegistrar)
+			{
```
Comment (spouliot) [inline tools/mmp/driver.cs]: style `{` should be on the same line for `if`

**Comment by chamons** [comment]
i386 is still broken, as Rolf determined via code inspection. I extended the xharness to generate/build 32-bit versions of our tests.

Still to do:

- Fix Code style issues ^
- Make tests pass
- Remove unnecessary 32 bit intro tests. classic and unified/32 and unified/64 as it only depends on what’s inside Xamarin.Mac.dll

File: tests/apitest/src/ObjCRuntime/SimpleRegistrarTest.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+#endif
+
+namespace Xamarin.Mac.Tests
+{
+	[Register ("RegistrarTestClass")]
+	class RegistrarTestClass : NSObject
+	{
+		public virtual string Value
+		{
```
Comment (rolfbjarne) [inline tests/apitest/src/ObjCRuntime/SimpleRegistrarTest.cs]: I think you need to configure your editor to put curly braces in the right place :smile:

File: src/notificationcenter.cs [hunk 11]
```diff
@@ -12,7 +12,7 @@
 
 namespace XamCore.NotificationCenter {
 #if XAMCORE_2_0 || !MONOMAC
-	[iOS (8,0)][Mac (10,10)]
+	[iOS (8,0)][Mac (10,10, onlyOn64 : true)]
```
Comment (rolfbjarne) [inline src/notificationcenter.cs]: The 32-bit test additions and the corresponding fixes are unrelated can go in a different PR.

Comment (chamons) [inline src/notificationcenter.cs]: Ugg. I know you are right but i didn't want to bother. Will do.

**Comment by chamons** [comment]
All tests changes have been ripped out and moved to #1234 

This PR now depends on that one landing first, merging in, and tests passing before landing.

**Comment by chamons** [comment]
Just merged in master so we'll get the 32-bit tests from that. 

If those come back clean then we can get final approval and land this finally.
